Introduce torch.distributed._shard package.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66459

Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

pritamdamania committed:
Introduce torch.distributed._shard package.

Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

bowangbj reviewed (approved):
Thanks Pritam for the great reorg. Looks pretty good!

wanchaol reviewed (approved):
looks great! Some windows CI failed on test_sharding_spec though, is there some changes of the testing before and after?

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Introduce torch.distributed._shard package."


Rearrange the _sharded_tensor and _sharding_spec packages to be under
a single path: torch.distributed._shard.

#Closes: https://github.com/pytorch/pytorch/issues/65628

Differential Revision: [D31559510](https://our.internmc.facebook.com/intern/diff/D31559510/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31559510/)!

[ghstack-poisoned]
